### DISCUSSION TITLE
Bucket4j with Apache Ignite

### DISCUSSION THREAD

**Discussion by KumariPrasanna** [vladimir-bukhtoyarov/bucket4j#237] [Q&A]
Hi Vladimir,
    I am trying to use bucket4j with Apache Ignite. My requirement is to have a separate bucket for each client. Therefore, depending on the client, I need to retrieve the appropriate bucket to check for available tokens. 

Here is the code which I have tried so far for getting the bucket. However, I am unclear on how to fetch client specific bucket. 

    publicBucket bucket(Ignite ignite) {
        Refill refill= Refill.greedy(capacity, java.time.Duration.ofMinutes(refillPeriod));
        Bandwidth limit= Bandwidth.classic(capacity, refill);
        IgniteCache<String, byte[]> igniteCache= ignite.getOrCreateCache(TOKEN_BUCKETS_CACHE_NAME);
        IgniteProxyManager<String> igniteProxyManager= newIgniteProxyManager<>(igniteCache);
        BucketConfiguration bucketConfiguration= BucketConfiguration.builder()
                                                                     .addLimit(limit)
                                                                     .build();
        return igniteProxyManager.builder().build("client1", bucketConfiguration);
 }

Let's say, first time I intercept a request based on which I create a client specific bucket. For example, if the client is "client1", I will create a "client1" bucket as described above ( similarly I will create "client2" bucket for "client2"). Where and how can I store this bucket so that I can retrieve the client specific bucket the next time when a request that needs to be rate-limited for this client is intercepted? Should I simply invoke " igniteProxyManager.builder().build("client1", bucketConfiguration);" to get the client specific bucket? Will this create a new bucket again or return an existing client bucket that was created the very first time. 

Thanks,
Prasanna

**Comment by vladimir-bukhtoyarov** [comment]
Hello @KumariPrasanna 

1. As described [in javadocs](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/RemoteBucketBuilder.java#L67) you are free to build as many buckets with the same key as you need, all proxies will point to the same bucket persisted in the storage. Buld->interact->forget is suitable sequence for most cases. So you do not have to store the references anywhere.
2. About code style - it is better to instantiate igniteProxyManager once and store it in a private field in order to avoid unnecessary memory allocations.

**Reply by KumariPrasanna** [reply]
Thanks @vladimir-bukhtoyarov for the replies!!!
